### ISSUE TITLE
element filters: how to only consider interfaces?

### ISSUE THREAD

**Issue by monperrus** [revapi/revapi#60]
In our project, I need that revapi only considers changes in public interfaces. How to do this?

(info not found at http://revapi.org/modules/revapi-basic-features/extensions/filter.html)

**Comment by metlos** [comment]
Interesting.. If you have all your interfaces grouped in a single package, you could filter by that. If you have your classes and interfaces mixed in a single package, I'm afraid this can't be done at the moment (unless you consider listing the interfaces one by one acceptable).

Or did you find Revapi looking for changes in non-public interfaces?

**Comment by monperrus** [comment]
> did you find Revapi looking for changes in non-public interfaces?

no, it's not a bug report, but more a feature request :-)

>  If you have all your interfaces grouped in a single package, you could filter by that.

not my case.

> unless you consider listing the interfaces one by one acceptable

not really a solution in a ten-year old code base with dozens and dozens of interfaces.

OK, so no solution for now. Thanks.

**Comment by metlos** [comment]
This is an interesting idea though. I have to think about how to best implement it. In a sense, we're looking at introducing something like a query language for java elements here... Today I've got something for looking up elements by their signature (sort of), but you'd like to expand that to include other "qualities" of the elements as well.

Coincidentally, `revapi-java` introduced something like this for difference matching in `0.13.0`. Unfortunately, revapi currently doesn't expose the kind of the element the difference deals with, but that is very easy to add.

So in `0.13.1` (that I will be releasing shortly, I think), you still won't be able to filter out elements from the analysis based on their kind (i.e. use `revapi.filter` configuration), but you will be able to ignore all differences for non-interface elements (the `elementKind` is the new matcher that I am adding for `0.13.1` based on your feature request):

```javascript
{
  "revapi": {
    "ignore": [
      {
        "regex": true,
        "code": ".*",
        "elementKind": "^(?!interface)$"
      }
    ]
  }
}
```
The above will ignore any kind difference that is found on any type type of element other than "interface".

**Comment by monperrus** [comment]
> So in |0.13.1| (that I will be releasing shortly, I think), you still
> won't be able to filter out elements from the analysis based on their
> kind (i.e. use |revapi.filter| configuration), but you will be able to
> ignore all differences for non-interface elements (the |elementKind|
> is the new matcher that I am adding for |0.13.1| based on your feature
> request):
Do you mean that the analysis will still be done but the output would
not contain the corresponding warnings?

**Comment by metlos** [comment]
> Do you mean that the analysis will still be done but the output would not contain the corresponding warnings?

Yes.

**Comment by monperrus** [comment]
Excellent! I look forward to the next release.

**Comment by metlos** [comment]
Implemented in `revapi-java-0.13.1`.

**Comment by metlos** [comment]
Also, this required a fix in `revapi-basic-features-0.5.1`. So in order to properly use this, you need to upgrade to `revapi-maven-plugin-0.8.1` that depends on that fixed version...

**Comment by monperrus** [comment]
thanks!
